scalanet/src/io/iohk/scalanet/peergroup/TLSPeerGroup.scala

```diff
@@ -240,23 +234,43 @@ object TLSPeerGroup {
                       s"Ssl Handshake client channel from ${ctx.channel().localAddress()} " +
                         s"to ${ctx.channel().remoteAddress()} with channel id ${ctx.channel().id} and ssl status ${e.isSuccess}"
                     )
-                    if (e.isSuccess) activation.success(ctx) else activation.failure(e.cause())
+                    if (e.isSuccess)
+                      activation.success(ctx)
+                    else {
+                      activation.failure(e.cause())
+                    }
 
                   case _ =>
                     log.debug(
                       s"User Event client channel from ${ctx.channel().localAddress()} " +
                         s"to ${ctx.channel().remoteAddress()} with channel id ${ctx.channel().id}"
                     )
                 }
-                super.userEventTriggered(ctx, evt)
               }
             })
             .addLast(new MessageNotifier[M](messageSubject, codec, bi))
         }
       })
 
+    private def mapException(t: Throwable): Throwable = t match {
+      case _: ClosedChannelException =>
+        new peergroup.PeerGroup.ChannelBrokenException[InetMultiAddress](to, t)
+      case _: ConnectException =>
+        new scalanet.peergroup.PeerGroup.ChannelSetupException[InetMultiAddress](to, t)
+      case _: SSLKeyException =>
+        new iohk.scalanet.peergroup.PeerGroup.HandshakeException[InetMultiAddress](to, t)
+      case _ =>
+        t
+    }
+
     def initialize: Task[ClientChannelImpl[M]] = {
-      toTask(bootstrap.connect(inetSocketAddress)).map(_ => this)
+      toTask(bootstrap.connect(inetSocketAddress))
+        .flatMap(_ => Task.fromFuture(activationF))
+        .map(_ => this)
+        .onErrorRecoverWith {
+          case t: Throwable =>
+            Task.raiseError(mapException(t))
```

Comment:
could this be re-written? 
In a way we have two futures that are converted into Tasks, then we are returning a Task with a reference to the object the client already used to call `initialise`, why couldn't we just return, something like Task.unit? avoiding the strange `map(_ => this)`
A related comment, isn't `activationF` evaluation running from the call to `activationF.future` (Line 210), what are we winning by adding it inside the Task that is discarding it anyway? Maybe I am misunderstanding Task semantics.

Reply:
Hmm, it is a bit of a mess, I agree. Had a play around and could not see anything much cleaner because there are 3 separate async constructs (netty's Future, scala's Future and Task). Think I will raise a story to upgrade monix (the next minor version picks up a bunch of new ways to create tasks) and then get rid of the scala Promises. Sound reasonable?

Reply:
It partly addresses the comment, two minor pending question are:
- Why return a reference to `this` inside the returned Task value? Couldn't you return Task[Unit]?
- Also, why involving the Task that calls the promise's future? My remark there was that it seems to be adding no purpose (as the future was already running and wrapping it inside the Task seems to add no effect).